keyserver/server.go

```diff
@@ -0,0 +1,224 @@
+	signPath := util.ResolvePath(conf.Policies.SignKeyPath, file)
+	signKey, err := ioutil.ReadFile(signPath)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot read signing key: %v", err)
+	}
+	if len(signKey) != sign.PrivateKeySize {
+		return nil, fmt.Errorf("Signing key must be 64 bytes (got %d)", len(signKey))
+	}
+
+	// load VRF key
+	vrfPath := util.ResolvePath(conf.Policies.VRFKeyPath, file)
+	vrfKey, err := ioutil.ReadFile(vrfPath)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot read VRF key: %v", err)
+	}
+	if len(vrfKey) != vrf.PrivateKeySize {
+		return nil, fmt.Errorf("VRF key must be 64 bytes (got %d)", len(vrfKey))
+	}
+
+	conf.configFilePath = file
+	conf.Policies.vrfKey = vrfKey
+	conf.Policies.signKey = signKey
+	// also update path for db & TLS cert files
+	conf.DatabasePath = util.ResolvePath(conf.DatabasePath, file)
+	conf.TLS.TLSCertPath = util.ResolvePath(conf.TLS.TLSCertPath, file)
+	conf.TLS.TLSKeyPath = util.ResolvePath(conf.TLS.TLSKeyPath, file)
+
+	return &conf, nil
+}
+
+func NewConiksServer(conf *ServerConfig) *ConiksServer {
+	// open db
+	kvdb := leveldbkv.OpenDB(conf.DatabasePath)
+
+	// create server instance
+	server := new(ConiksServer)
+	server.dir = protocol.InitDirectory(
+		conf.Policies.EpochDeadline,
+		conf.Policies.vrfKey,
+		conf.Policies.signKey,
+		conf.LoadedHistoryLength,
+		true)
+	server.stop = make(chan struct{})
+	server.db = kvdb
+	server.configFilePath = conf.configFilePath
+	server.reloadChan = make(chan os.Signal, 1)
+	signal.Notify(server.reloadChan, syscall.SIGUSR2)
+	server.epochTimer = time.NewTimer(time.Duration(conf.Policies.EpochDeadline) * time.Second)
+
+	return server
+}
+
+func (server *ConiksServer) Run(tc *TLSConnection) {
+	server.waitStop.Add(1)
+	go func() {
+		server.EpochUpdate()
+		server.waitStop.Done()
+	}()
+
+	// server listener
+	cer, err := tls.LoadX509KeyPair(tc.TLSCertPath, tc.TLSKeyPath)
+	if err != nil {
+		panic(err)
+	}
+
+	server.tlsConfig = &tls.Config{Certificates: []tls.Certificate{cer}}
+
+	addr, err := net.ResolveTCPAddr("tcp", tc.PublicAddress)
+	if err != nil {
+		panic(err)
+	}
+
+	// server public port
+	publicLn, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		panic(err)
+	}
+
+	addr, err = net.ResolveTCPAddr("tcp", tc.LocalAddress)
+	if err != nil {
+		panic(err)
+	}
+
+	// server registration port
+	localLn, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		panic(err)
+	}
+
+	// acceptable types for public connection
+	publicTypes := make(map[int]bool)
+	server.waitStop.Add(1)
+	go func() {
+		server.listenForRequests(publicLn, server.makeHandler(publicTypes))
+		server.waitStop.Done()
+	}()
+
+	// acceptable types for local connection
+	localTypes := make(map[int]bool)
+	localTypes[protocol.RegistrationType] = true
+	server.waitStop.Add(1)
+	go func() {
+		server.listenForRequests(localLn, server.makeHandler(localTypes))
+		server.waitStop.Done()
+	}()
+
+	server.waitStop.Add(1)
+	go func() {
+		server.updatePolicies()
```

Comment:
I think I can explain a little bit here. The server's `reloadChan` will listen for a signal from the user (`kill -USR2 pid`), so what `updatePolicies()` does is to receive this signal and re-read the policies from the config file.